### DISCUSSION TITLE
Shell to Windows Using DOSBox-X

### DISCUSSION THREAD

**Discussion by Ujammer** [joncampbell123/dosbox-x#4341] [Q&A]
I have some DOS Apps that also SHELL OUT to Windows Apps. Is it possible using DOSBOX-X to SHELL OUT a windows app? For example, I have line of code in my DOS App, **SHELL "f:\MDATE.EXE"**. DOXBOX-X executes the code, but I get the error, "This program cannot be run in DOS mode." which is the standard error you would get trying to run a Windows executable in DOS.

If I run the code on an older box that is running native MS-DOS, the code works fine and it shells out to the Windows program with no errors. 

I've looked in the Wiki for a solution, but I didn't find an answer there. Does anyone have an idea if this will work on DOSBOX-X? Thanks in advance for your help!

Update: I got this to work by changing the DOSBox-X .confg file. I changed the version (ver) of DOS to 7.1, set startcmd = true, and startquiet = true. It will now shell out to windows programs and run them. Thanks for all the responses!!

**Comment by RBTCHENTPE** [comment]
Dear Sir,

 

Is it possible to set the dosbox-x.conf for the function requirement as follows?

 

Require : Auto-mount drives of my Windows drive “F:” (makes DOSBox-X access to the same drives as in Windows system) which drives is linked (map) to network driver when the 

DOSBOX-S started.

 

Refer to the instruction :

If you are using Windows, DOSBox-X will ask if you want to give it access to your Windows drive when you try to go to a drive (e.g. C:), but that drive has not yet been mounted inside DOSBox-X. 

This basically makes DOSBox-X access to the same drives as in your Windows system. If you answer Y for Yes, then the whole Windows drive will be mounted and accessible within DOSBox-X. 

You can also automatically mount drives via the "Drive" menu (select a drive and "Mount Automatically").

Furthermore, it is possible to automatically mount all available Windows drives as DOSBox-X drives at start by setting the "automountall" config option to "true". Note however that it is not recommended mounting your whole Windows Drive C: inside DOSBox-X because DOSBox-X will then be able to access all files and directories in your Windows Drive C:, which may not be very safe.

 

B. Rgds.,

Robert Chen

 

 

From: EugeneHenson ***@***.*** 
Sent: Saturday, June 17, 2023 5:50 AM
To: joncampbell123/dosbox-x
Cc: Subscribed
Subject: [joncampbell123/dosbox-x] Shell to Windows Using DOSBox-X (Discussion #4341)

 

I have some DOS Apps that also SHELL OUT to Windows Apps. Is it possible using DOSBOX-X to SHELL OUT a windows app? For example, I have line of code in my DOS App, SHELL "f:\MDATE.EXE". DOXBOX-X executes the code, but I get the error, "This program cannot be run in DOS mode." which is the standard error you would get trying to run a Windows executable in DOS.

If I run the code on an older box that is running native MS-DOS, the code works fine and it shells out to the Windows program with no errors.

I've looked in the Wiki for a solution, but I didn't find an answer there. Does anyone have an idea if this will work on DOSBOX-X? Thanks in advance for your help!

**Comment by Torinde** [comment]
You can install in DOSbox-X an operating system and boot it: MS-DOS, Windows 9x.
Is your program some kind of Windows NT console program?
Maybe the best is just to share the name (or attach the ZIP) of the program.

**Reply by Ujammer** [reply]
I wrote a simple Quick Basic test program in DOS that has one command "SHELL f:\je\MDATE.exe". I made a DOS executable file from that program and put it in the f:\je folder. the program is called tstshell.exe. 

MDATE.exe is a simple Windows app written in .net that takes the current date and calculates the day of the year. For example, today's date of 6/19/23 calculates to the 170th day of this year. 

If everything is working correctly when I run tstshell.exe in DOSBox-X, it should shell out to to the windows program MDATE.exe and run it in Windows mode. Instead it throws an error that says, "This program cannot be run in **DOS mode**." 

So it appears that DOSBox-X is trying to SHELL OUT to the Windows program MDATE, but it just isn't running it in Windows mode. 

I believe that I have mounted all the drives in DOXBox-X necessary to make this work. I'm running in Windows 10.

**Reply by Torinde** [reply]
what is this "SHELL"? a Quick Basic command?
From what you say I can't see how you expect it to work. Quick Basic is a DOS program from long before .NET, so even if Quick Basic supports some kind of Win3.x trickery (when run from a DOS prompt??) - it definitely doesn't support .NET and it definitely doesn't support "breaking out to host Windows" to run .NET programs in Win10. Quick Basic is UNAWARE of being run in an emulator (DOSbox) inside Windows 10 - from QuickBasic perspective it runs on a "Computer with DOS".

If I understand correctly what you're trying to do - maybe make a script on the Win10 side, which runs this .NET MDATE.EXE, gets the output in a parameter and passes that parameter to DOSbox somehow (via .conf file or some command line parameter)??

**Comment by emendelson** [comment]
You can do this by using a command like this, entered at the DOSBox prompt, and it will probably also run from your SHELL command:

`command /c start c:\windows\write.exe`
